Modules/feature/sh 198 automatic change ip

KarasinskiKacper opened the pull request:
Add:
- namespace "Comms" for communication related features
- Connection class for handling connection
- automatic change of IP and RF channel
- documentation for namespaces

Change:
- some special messages
- moved "repeat" logic to Connection class

KarasinskiKacper committed:
SH-198 Add for module methods checking IP and MAC addresses

KarasinskiKacper committed:
SH-197 Add for central unit methods checking IP and MAC addresses

KarasinskiKacper committed:
SH-197 Add Connection classes

KarasinskiKacper committed:
SH-198 Add timer, queues and mutex to Connection class

for:
- connection request timeout
- messages
- mAckNumber

KarasinskiKacper committed:
SH-198 Update handling messages in Connection class

KarasinskiKacper committed:
SH-198 Update special messages, add connection handling in end of decode

KarasinskiKacper committed:
SH-198 Resolve TODOs

KarasinskiKacper committed:
SH-198 Remove deliver classes of Connection class, add changeRFChannel()

KarasinskiKacper committed:
SH-198 Add method for setting protocol IP for central unit

KarasinskiKacper committed:
SH-198 Add automatic changing rf channel, NOT TESTED

KarasinskiKacper committed:
SH-198 Add handling of new connection, change special message ack to 255

KarasinskiKacper committed:
SH-198 Rollback TCP connection - part 1 NOT WORKING

KarasinskiKacper committed:
SH-198 Rollback TCP connection - part 2 fix addressing algorithm

KarasinskiKacper committed:
SH-198 Rollback TCP connection - part 3 NOT WORKING

Add automatic changing IP and rf Channel

KarasinskiKacper committed:
SH-198 Change ERROR and WARMING logs

KarasinskiKacper committed:
SH-198 Rollback TCP connection - part 4 (final)

fix problems with automatic changing rf channels

KarasinskiKacper committed:
SH-198 Add documentation for namespaces

KarasinskiKacper committed:
Revert "SH-198 Add documentation for namespaces"

This reverts commit 537c6bb469e7661c073f1df3afdabebf04e5cf9b.

KarasinskiKacper committed:
SH-198 Add documentation for namespaces

KarasinskiKacper committed:
SH-198 Cleanup code, add comments (not for Connection class)

KarasinskiKacper committed:
SH-198 Add Connection class comments, final cleanup before pull request

KarasinskiKacper committed:
SH-198 Fix missed errors in comments, before pull request

DyczekPawel commented on modules/src/communication/addressing/addressing.h:
```diff
@@ -9,156 +9,198 @@
-    void startAddressing();
+        /**
+         * @brief Copies the current protocol MAC address into the provided buffer.
+         * @param macAddress Output buffer of length 6 for the MAC address.
```
```suggestion
         * @param macAddress Output buffer of length defined by MAC_ADDRESS_LENGTH for the MAC address.
```

DyczekPawel commented on modules/src/communication/addressing/addressing.h:
```diff
@@ -9,156 +9,198 @@
+         * @return True if IP address is proper, false otherwise.
+         * @warning Must be implemented by derived class.
+         */
+        virtual bool isIpProper(uint8_t ip) = 0;
```
Valid zamiast Proper?

DyczekPawel commented on modules/src/communication/addressing/addressing.h:
Sugestia do porządkowania funkcji, może lepiej porządkować je w taki sposób:
1. virtual = 0
2. virtual
3. normal

W taki sposób w jednym miejscu masz wszystkie funkcje do zaimplementowania.

DyczekPawel commented on modules/src/communication/addressing/addressing.cpp:
```diff
@@ -12,128 +12,137 @@
-    if (mAddressingTaskHandle != nullptr) {
-        vTaskDelete(mAddressingTaskHandle);
-        mAddressingTaskHandle = nullptr;
+    bool Addressing::getIsAddressingWorking() const {
```
Running lub InProgress zamiast Working? Bardziej sygnalizuje że aktywnie się wykonuje, to samo do zmiennej.

DyczekPawel commented on modules/src/communication/addressing/addressing.cpp:
```diff
@@ -12,128 +12,137 @@
+
+        xSemaphoreTake(mAddressingDataMutex, portMAX_DELAY);
+        mIsAddressingWorking = false;
+        xSemaphoreGive(mAddressingDataMutex);
```
Użycie typu atomic dla mIsAddressingWorking zamiast mutexów.

DyczekPawel commented on modules/src/communication/addressing/central_unit_addressing.cpp:
```diff
@@ -13,393 +13,486 @@
+
+    uint8_t CentralUnitAddressing::getIPAddress() {
+        uint8_t result;
+        xSemaphoreTake(mModulesAddressingDataMutex, portMAX_DELAY);
+        if (mTmpModuleIp == NULL_IP || mIsStartOfAddressing) {
+            result = CENTRAL_UNIT_IP;
+        } else {
+            result = mTmpModuleIp;
+        }
+        xSemaphoreGive(mModulesAddressingDataMutex);
```
Zmiana typu zmiennych na atomic dla pozbycia się mutexów.

KarasinskiKacper replied:
W tym miejscu akurat zostaje mutex, ze względu na synchronizację tasków.

DyczekPawel commented on modules/src/communication/addressing/central_unit_addressing.cpp:
```diff
@@ -13,393 +13,486 @@
+        }
+    }
+
+    bool CentralUnitAddressing::isMACProper(const uint8_t *mac) {
```
Valid zamiast proper?

DyczekPawel commented on modules/src/communication/addressing/central_unit_addressing.cpp:
```diff
@@ -13,393 +13,486 @@
+        return result;
+    }
+
+    bool CentralUnitAddressing::isIpProper(const uint8_t ip) {
```
Valid zamiast proper?

DyczekPawel commented on modules/src/communication/addressing/central_unit_addressing.cpp:
```diff
@@ -13,393 +13,486 @@
+
+    bool CentralUnitAddressing::isIpProper(const uint8_t ip) {
+        bool result = true;
+        if ((getIsStartOfAddressing() && ip != NULL_IP) || (!getIsStartOfAddressing() && (ip == CENTRAL_UNIT_IP || ip == NULL_IP))) {
```
Złożony warunek, może warto dodać komentarz.

DyczekPawel commented on modules/src/communication/addressing/central_unit_addressing.cpp:
```diff
@@ -13,393 +13,486 @@
-    return tmpModuleIp;
-} 
+    void CentralUnitAddressing::setTmpModuleIp(const uint8_t ip) {
+        xSemaphoreTake(mModulesAddressingDataMutex, portMAX_DELAY);
+        mTmpModuleIp = ip;
+        xSemaphoreGive(mModulesAddressingDataMutex);
```
Zmiana typu zmiennej na atomic.

KarasinskiKacper replied:
tak samo jak tutaj https://github.com/Projekt-Inzynierski-Dyczek-Karasinski/smart-home/pull/11#discussion_r2384097172

DyczekPawel commented on modules/src/communication/addressing/central_unit_addressing.cpp:
```diff
@@ -13,393 +13,486 @@
-}
-// ================================================================
\ No newline at end of file
+    bool CentralUnitAddressing::getIsStartOfAddressing() const {
+        xSemaphoreTake(mModulesAddressingDataMutex, portMAX_DELAY);
+        const bool result = mIsStartOfAddressing;
+        xSemaphoreGive(mModulesAddressingDataMutex);
+        return result;
+    }
+
+    void CentralUnitAddressing::setIsStartOfAddressing(const bool value) {
+        xSemaphoreTake(mModulesAddressingDataMutex, portMAX_DELAY);
+        mIsStartOfAddressing = value;
+        xSemaphoreGive(mModulesAddressingDataMutex);
+    }
```
Zmiana typu mIsStartOfAddressing na atomic.

KarasinskiKacper replied:
To zostaje ze względu na synchronizację dostępu do wielu zmiennych

DyczekPawel commented on modules/src/communication/addressing/central_unit_addressing.h:
```diff
@@ -12,135 +12,206 @@
+         * @return True if IP address is proper, false otherwise.
+         * @note Thread-safe.
+         */
+        bool isIpProper(uint8_t ip) override;
```
Zmiana proper na valid (funkcje i docstring).

DyczekPawel commented on modules/src/communication/addressing/module_addressing.cpp:
Dotyczą się komentarze z central_unit_addressing.cpp do wprowadzenia typów atomic i zmiany Proper na Valid.

DyczekPawel commented on modules/src/communication/addressing/module_addressing.h:
Dotyczą się komentarze z central_unit_addressing.h do zmiany Proper na Valid.

DyczekPawel commented on modules/src/communication/connection/connection.cpp:
```diff
@@ -0,0 +1,196 @@
+        } else if (uah::areArraysEqual(receivedMessage, (uint8_t*)CONNECTION_RE_TEST_GET, 26)) {
+            uah::prepareBuffer(sendBuffer, (uint8_t*)CONNECTION_END, 5, MESSAGE_SIZE);
+            mpCommunication->sendMessage(sendBuffer);
+            endConnection();
```
Czy ręczne podawanie długości bufora nie mogło by być zastąpione użyciem strlen().
```
size_t size = strlen(CONNECTION_RE_TEST_GET);
```
Ta uwaga dotyczy się wielu plików.

DyczekPawel commented on modules/src/communication/communication.cpp:
```diff
@@ -5,815 +5,789 @@
+                            isRawMessage = false;
+                            com.mpLogger->infoa("Communication Decode", "Received raw message: ", protocolBuffer[0], PROTOCOL_SIZE, false);
+                            xQueueSend(com.mReceiveMessageQueue, protocolBuffer[0], portMAX_DELAY);
+                        } else if (uah::calcLenOfDataInArray(messageBuffer, 2) != 1) {
```
Może komentarz? `!= 1` jest troche mało intuicyjne, jeżeli to ma na celu zabezpieczyć przed wysłaniem pustej wiadomości to może lepiej użyć warunku `> 1`

KarasinskiKacper replied:
Ten warunek to pozostałość po usuniętej funkcjonalności - usunąłem go.

DyczekPawel reviewed (changes requested):
Możesz rozważyć poprawienie nazw zmiennych/funkcji (z uwzględnieniem docstringów) oraz wprowadzenie zmiennych atomic.  
Dodatkowe sugestie: 
- wyciągnięcie folderu docs poziom wyżej, żeby był równoległy do src
- wprowadzenie zmiennych przechowujących ilość przydzielonej pamięci do tasków rtos (chyba byłoby wygodniej mieć wszystkie wartości pamięci dla tasków w jednym miejscu np: smart_home_config.h)

KarasinskiKacper committed:
SH-198 Add macros for FreeRTOS task sizes, move folder docs

KarasinskiKacper committed:
SH-198 Update comments and method names, add overloads in ArrayHandles

Update comments and method names for addressing classes (proper -> valid), add comment for if requested in pull review, add overloads for prepareBuffer and areArraysEqual for char arrays

KarasinskiKacper committed:
SH-198 Move virtual methods to top (Addressing class)

KarasinskiKacper committed:
SH-198 Change some class scope vars to atomic type
